File: y2017/rust/src/bin/day1.rs
use std::fs;

fn main() {
    let mut digits = fs::read_to_string("input/day1.txt").expect("bruh");
    digits = digits.trim().parse().unwrap();

    let digit_vec: Vec<char> = digits.chars().collect();

    let mut p1_total = 0;
    let mut p2_total = 0;
    for i in 0..digits.len() {
        let next1 = (i + 1) % digits.len();
        if digit_vec[i] == digit_vec[next1] {
            p1_total += digit_vec[i].to_digit(10).unwrap_or(0);
        }

        let next2 = (i + digits.len() / 2) % digits.len();
        if digit_vec[i] == digit_vec[next2] {
            p2_total += digit_vec[i].to_digit(10).unwrap_or(0);
        }
    }

    println!("solution to p1 captcha: {p1_total}");
    println!("solution to p2 captcha: {p2_total}");
}

File: y2017/nim/day1.nim
import strutils

let captcha = readFile("input/day1.txt").strip()
assert captcha.len() mod 2 == 0, "captcha length should be even"

let cLen = captcha.len()
var
  p1Tot = 0
  p2Tot = 0
for i in 0 ..< cLen:
  let currVal = int(captcha[i]) - int('0')
  let next1 = captcha[(i + 1) mod cLen]
  let next2 = captcha[(i + cLen div 2) mod cLen]
  p1Tot += int(captcha[i] == next1) * currVal
  p2Tot += int(captcha[i] == next2) * currVal

echo "solution to p1 captcha: $1" % [$p1Tot]
echo "solution to p2 captcha: $1" % [$p2Tot]
